rfc/java11.md

```diff
@@ -0,0 +1,164 @@
+before, the `java8` variants will offer a way to opt out of this change.
+
+Other than that, everything from the previous section applies to this release as
+well.
+
+### 1.20 (early May)
+
+For this release, we will do two things:
+
+1.  Compile with JDK 11 using the `-source 8 -target 8` flags. This produces
+    builds that are still compatible with Java 8, and means the build will still
+    fail if anyone tries to use Java 11 APIs or language features.
+2.  Remove the `java8` container variants
+
+If the compiler change causes issues, we can revert that change and push that
+out to the next release.
+
+If we know for a fact (from previous discussions) that removing the `java8`
+variants will cause problems for some customers, or if a customer complains
+after the release, we can put back the `java8` containers and delay this part
+(but keep the compiler change) until the next release.
+
+### 1.21 (early July)
+
+Change `-source 8 -target 8` to `-source 11 -target11`. At this point,
+developers are able to commit Java 11 code to the repositories. Undoing this
+change is possible, but will require some work. As such, this should only be
+done after
+
+1.  the `java8` containers have been removed for a while without complaint
+2.  we don't know of any other non-container customers still running on Java 8
+
+## Prior Art and Alternatives
+
+The general approach was modelled after how Google migrates its monorepo to new
+Java versions:
+
+1.  switch the JRE to the new version
+2.  use the new JDK's `javac` with `-target 8 -source 8`
+3.  remove those flags
+
+Between each step is a period of time to evaluate breakages and potentially roll
+back.
+
+## Known Unknowns
+
+On the JRE side, there are probably changes to performance, memory usage, and
+anything else that could be affected by a major runtime change like this.
+
+On the JDK side, it's not clear what amount of code changes will be required for
+Java 11 compatibility. It's likely that certain libraries won't work perfectly
+with Java 11, for example, or might require a major version upgrade to do so.
+(My very brief initial testing quickly found some issues with
+[Spock](http://spockframework.org/), for example.)
+
+My hope is that the very slow rollout will help shake out these issues, and
+we'll relax the timeline as needed.
+
+## Operations
+
+Individual operations teams who aren't using the containers will need to upgrade
+their runtimes to Java 11 between now and (tentatively) Spinnaker 1.21. My
+assumption is that companies that aren't using the container builds also have
+the in-house Spinnaker and Java expertise to handle this runtime change without
+too many problems.
+
+## Future Possibilities
+
+By the time this RFC is finished, Java 14 will have been released. This is the
+next LTS release of Java. This RFC should be updated with any changes to the
+plan so that it can be used as a template for the Java 14 upgrade, whenever we
+decide that should happen.
```

Comment:
Are we hoping that, going forward, we continue to track more closely with LTS as they're release or handle this on a case-by-case basis?

Reply:
I think we have to handle this case-by-case. We can't introduce a dependency on Java 14 until Netflix is comfortable running Java 14, for example. So it's going to have to be dependent on our customers being ready to run the new JREs.

That said, the LTS releases should be closer together now, so the diff between Java 11→14 is much smaller than 8→11.

Reply:
I would just add that upgrading for the sake of upgrading is probably not worth the risk unless we are getting something of benefit from doing so

Reply:
I generally agree with that, although we do have to upgrade at some point just to keep up with the Java ecosystem, regardless of whether or not we directly are benefiting. At some point libraries we depend on will stop supporting old JREs because they're compiled with a newer JDK, etc. Upgrading too early has its own risks, but I'd also personally prefer to be a little proactive so we don't end up in a situation where our hand is forced because we need to upgrade some important library and can't until we go through the six month JDK upgrade process detailed here...